0000-zmix/README.md

```diff
@@ -0,0 +1,200 @@
+multiple credentials that can fulfill the proof. The choices were made will need to be communicated to the verifier as a **proof resolution**–the non-cryptographic
+description about the proof. The **proof resolution** and **proof request** can be combined with the public data to form the **proof spec**.
+Exactly where the public cryptographic material is stored and how it is retrieved is up to end consumers.
+This could be handled later.
+
+## Example diagram
+
+![flow](zmix_flow.png)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+There are six ZKPs that ZMix can do:
+
+1. *Commitments* - Proof of knowledge of a secret value
+    - Required public parameters: generators, modulus/curve
+1. *Scoped Commitments* - Like *commitments* but restricted to a specific context.
+    - Required public parameters: Hash, context label, map to resulting generator, modulus
+1. *Signature Proofs of Knowledge* - Instead of revealing the signature, prove knowledge of it, allows proving knowledge of messages included in the signature. Messages can be hidden or revealed.
+    - Required public parameters: schemas, encodings, public keys, curves/modulus
+1. *Intervals* - Proof a value lies between a lower and upper bound
+    - Required public parameters:
+        - RSA: generators and modulus
+        - ECC: Bulletproofs or R1CS
+        - Circuit logic: public equations
+1. *Set Memberships* - Proof a value is in a set of known values
+    - Required public parameters:
+        - Static/Enum: Set of known values
+        - Dynamic: Accumulator with public parameters or Merkle tree with circuit logic
+1. *Verifiable Encryption* - Proof a third party knows a secret that can decrypt a value
+    - Required public parameters: generators, modulus/curve, public key, relation description
+
+To produce a proof, a proof spec and a witness are inputs for the ZMix create proof function.
+To verify a proof, a proof spec and a proof are inputs for the ZMix verify proof function.
+
+The ZMix parser should map the subproofs to their appropriate crypto primitives
+provided either directly in ZMix or in libursa.
+
+The proof spec format is described [here](proof-spec.json)
+The witness format is described [here](witness.json)
+The proof format is described [here](proof.json)
+
+ZMix will provide objects for composing proofs directly in Rust or serializing
+from JSON.
+
+The following code is proposed
+
+```rust
+#[derive(Serialize, Deserialize)]
+pub enum ProofSpecClause {
+    Credential(...),
+    Interval(...),
+    SetMembership(...),
+    VerifiableEncryption(...),
+    Commitment(...),
+    ScopeCommitment(...)
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProofSpec {
+    pub clauses: HashMap<String, ProofSpecClause>
+}
+
+impl ProofSpec {
+    pub fn new() -> ProofSpec {
+        ProofSpec {
+            clauses: HashMap::new()
+        }
+    }
+}
+
```

Comment:
I like this modular structure. Can we verify that it will handle (almost) everything we want to implement in the future?